File: WebStorm/JavaScript/eruru.mvvm/MVVMControl.js
function MVVMControl (control) {

	this.root = null;
	this.control = control;
	this.bindings = [];
	this.controlNames = [];
	this.controls = [];
	this.name = null;
	this.control.mvvmcontrol = this;
	this.itemsControlControls = [];
	this.mvvmPropertyChanged = new MVVMEvent ();

	this.registerControl = function (control) {
		control.parent = this;
		control.root = this.root;
	};

	this.add = function (control) {
		this.insert (this.controls.length, control);
	};

	this.insert = function (index, control) {
		this.registerControl (control);
		this.controls.splice (index, 0, control);
	};

	this.removeAt = function (index) {
		this.controls.splice (index, 1);
		this.controls[index].unbind ();
	};

	this.move = function (oldIndex, newIndex) {
		var control = this.controls[oldIndex];
		this.controls.splice (oldIndex, 1);
		this.controls.splice (newIndex, 0, control);
	};

	this.set = function (index, control) {
		this.controls[index].unbind ();
		this.controls[index] = control;
		this.registerControl (control);
	};

	this.clear = function () {
		for (var i = 0; i < this.controls.length; i++) {
			this.controls[i].unbind ();
		}
		this.controls.splice (0);
	};

	this.setBinding = function (binding, setTargetValue, unbind, isDataContext, propertyName) {
		if (!isDataContext && !binding) {
			this.bindings.push (binding);
		}
		if (binding == null) {
			return null;
		}
		this[propertyName] = binding;
		binding.control = this;
		binding.onUnbind = unbind;
		binding.onSetTargetValue = setTargetValue;
		binding.targetPropertyName = propertyName;
		binding.isDataContext = isDataContext;
		binding.defaultMode = MVVMBindingMode.oneWay;
		binding.defaultUpdateSourceTrigger = MVVMBindingUpdateSourceTrigger.propertyChanged;
		binding.autoSetType ();
		var control = this;
		switch (propertyName) {
			case "value":
				this.control.addEventListener ("input", function () {
					this.mvvmcontrol.onChanged (this.mvvmcontrol.value, MVVMBindingOnChangedType.propertyChanged);
				});
				this.control.addEventListener ("blur", function () {
					this.mvvmcontrol.onChanged (this.mvvmcontrol.value, MVVMBindingOnChangedType.lostFocus);
				});
				binding.defaultMode = MVVMBindingMode.twoWay;
				binding.defaultUpdateSourceTrigger = MVVMBindingUpdateSourceTrigger.lostFocus;
				break;
			case "class":
				binding.blockSetTargetValue = true;
				binding.onSetTargetValue = function (targetValue) {
					if (typeof targetValue === "boolean") {
						if (targetValue) {
							control.control.classList.add (binding.sourcePropertyName);
						} else {
							control.control.classList.remove (binding.sourcePropertyName);
						}
						return;
					}
					control.control.className = targetValue;
				};
				break;
			case "itemssource":
				binding.blockSetTargetValue = true;
				this.itemTemplate = mvvmFindByAttribute (this.control, "mvvmitemtemplate");
				this.itemTemplate.style.display = "none";
				binding.onSetTargetValue = function (targetValue) {
					control.cancelCollectionChanged ();
					control.itemsControlReset ();
					if (targetValue != null && targetValue.forEach) {
						targetValue.forEach (function (value) {
							control.itemsControlAdd (value);
						});
					}
					if (targetValue instanceof MVVMObservableCollection) {
						this.notifyCollectionChanged = targetValue;
						this.notifyCollectionChanged.collectionChanged.add (this, control.collectionChanged);
					}
				};
				binding.onUnbind = this.cancelCollectionChanged;
				break;
			default:
				if (mvvmStringStartsWith (propertyName, "on")) {
					binding.blockSetTargetValue = true;
					this.control.addEventListener (propertyName.substring (2), function () {
						this.mvvmcontrol.onCommand (this.mvvmcontrol[propertyName], this.mvvmcontrol[propertyName + "parameter"]);
					});
				}
				break;
		}
		if (!isDataContext) {
			this.bindings.push (binding);
		}
		return binding;
	};

	this.setDataContext = function (binding) {
		binding.blockSetTargetValue = true;
		this.setBinding (binding, function (targetValue) {
			for (var i = 0; i < this.bindings.length; i++) {
				this.bindings[i].bind ();
			}
			for (i = 0; i < this.controls.length; i++) {
				this.controls[i].getDataContext ().bind ();
			}
		}, null, true, "dataContext");
	};

	this.getDataContext = function () {
		if (!this.hasOwnProperty ("dataContext")) {
			this.setDataContext (new MVVMBinding ());
		}
		return this.dataContext;
	}

	this.build = function (root) {
		if (root == null) {
			root = this.root == null ? this : this.root;
		}
		this.controlNames = [];
		this.unbind ();
		this.forEach (function (control) {
			control.register (root);
		});
		this.getDataContext ().bind ();
		return this;
	};

	this.forEach = function (action) {
		action (this);
		for (var i = 0; i < this.controls.length; i++) {
			this.controls[i].forEach (action);
		}
	};

	this.register = function (root) {
		this.root = root;
		if (this.name != null) {
			root.controlNames.put (this.name.getTargetValue (), this);
		}
	};

	this.unbind = function () {
		this.forEach (function (control) {
			control.getDataContext ().unbind ();
			for (var i = 0; i < control.bindings.length; i++) {
				control.bindings[i].unbind ();
			}
		});
	};

	this.onChanged = function (binding, onChangedType) {
		if (binding == null || binding.blockOnChanged) {
			return;
		}
		switch (binding.getMode ()) {
			case MVVMBindingMode.twoWay:
			case MVVMBindingMode.oneWayToSource:
				switch (binding.getUpdateSourceTrigger ()) {
					case MVVMBindingUpdateSourceTrigger.propertyChanged:
						if (onChangedType === MVVMBindingOnChangedType.propertyChanged) {
							binding.oneWaySetSourceValue ();
						}
						break;
					case MVVMBindingUpdateSourceTrigger.lostFocus:
						if (onChangedType === MVVMBindingOnChangedType.lostFocus) {
							binding.oneWaySetSourceValue ();
						}
						break;
				}
				break;
		}
		this.mvvmPropertyChanged.invoke (this, binding.targetPropertyName);
	};

	this.onCommand = function (command, parameter) {
		if (command == null) {
			return;
		}
		var relayCommand = command.getTargetValue ();
		if (relayCommand == null) {
			return;
		}
		var value = parameter == null ? null : parameter.getTargetValue ();
		if (relayCommand instanceof Function) {
			relayCommand.call (command.instance, value);
			return;
		}
		if (relayCommand.canExecute (value)) {
			relayCommand.execute (value);
		}
	}

	this.cancelCollectionChanged = function () {
		if (this.notifyCollectionChanged != null) {
			this.notifyCollectionChanged.collectionChanged.remove (this);
			this.notifyCollectionChanged = null;
		}
	}

	this.collectionChanged = function (sender, action, oldIndex, oldItems, newIndex, newItems) {
		switch (action) {
			case MVVMCollectionChangedAction.add:
				this.itemsControlInsert (newItems[0], newIndex);
				break;
			case MVVMCollectionChangedAction.remove:
				this.itemsControlRemoveAt (oldIndex);
				break;
			case MVVMCollectionChangedAction.replace:
				this.itemsControlReplace (oldIndex, newItems[0]);
				break;
			case MVVMCollectionChangedAction.move:
				this.itemsControlMove (oldIndex, newIndex);
				break;
			case MVVMCollectionChangedAction.reset:
				this.itemsControlReset ();
				break;
			default:
				throw "未实现" + action;
		}
	}

	this.itemsControlConvert = function (value) {
		var node = this.itemTemplate.cloneNode (true);
		node.removeAttribute ("mvvmitemtemplate");
		node.style.display = "";
		return mvvm.compile (node);
	};

	this.itemsControlRegister = function (value, index) {
		var control = this.itemsControlConvert (value);
		control.parent = this;
		control.root = this.root;
		control.setDataContext (new MVVMValueBinding (value));
		control.build (this.root);
		this.itemsControlControls.splice (index, 0, control);
		var node = this.itemTemplate;
		for (var i = 0; i <= index; i++) {
			node = node.nextSibling;
		}
		this.itemTemplate.parentNode.insertBefore (control.control, node);
	};

	this.itemsControlAdd = function (value) {
		this.itemsControlInsert (value, this.itemsControlControls.length);
	};

	this.itemsControlInsert = function (value, index) {
		this.itemsControlRegister (value, index);
	};

	this.itemsControlRemoveAt = function (index) {
		this.itemTemplate.parentNode.removeChild (this.itemsControlControls[index].control);
		this.itemsControlControls[index].unbind ();
		this.itemsControlControls.splice (index, 1);
	};

	this.itemsControlReplace = function (index, value) {
		this.itemsControlRemoveAt (index);
		this.itemsControlInsert (value, index);
	};

	this.itemsControlMove = function (oldIndex, newIndex) {

	};

	this.itemsControlReset = function () {
		for (var i = this.itemsControlControls.length - 1; i >= 0; i--) {
			this.itemsControlRemoveAt (i);
		}
	};

}

File: Unity/Assets/Eruru.MVVM/Scripts/Shared/Controls/MVVMControl.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace Eruru.MVVM {

	public partial class MVVMControl : INotifyPropertyChanged, IEnumerable<MVVMControl> {

		public event PropertyChangedEventHandler PropertyChanged;
		public MVVMBinding Name {

			get {
				return GetBinding (ref _Name, binding => Name = binding);
			}

			set {
				SetBinding (ref _Name, value);
			}

		}
		public MVVMBinding DataContext {

			get {
				return GetBinding (ref _DataContext, binding => {
					binding.Path = null;
					DataContext = binding;
				});
			}

			set {
				SetBinding (ref _DataContext, value, isDataContext: true, setTargetValue: targetValue => {
					foreach (MVVMBinding binding in Bindings) {
						binding.Bind ();
					}
					foreach (MVVMTrigger trigger in Triggers) {
						trigger.Bind ();
					}
					foreach (MVVMControl control in Controls) {
						control.DataContext.Bind ();
					}
				});
			}

		}
		public MVVMControl Root {

			get {
				return _Root;
			}

		}
		public MVVMControl Parent {

			get {
				return _Parent;
			}

		}
		public List<MVVMControl> Controls {

			get {
				return _Controls;
			}

		}
		public MVVMControl this[int index] {

			get {
				return Controls[index];
			}

			set {
				Controls[index].Unbind ();
				Controls[index] = value;
			}

		}
		public MVVMValidation Validation {

			get {
				return _Validation;
			}

			set {
				_Validation = value;
			}

		}

		internal Dictionary<string, MVVMControl> ControlNames = new Dictionary<string, MVVMControl> ();
		internal MVVMControl _Parent;

		readonly List<MVVMBinding> Bindings = new List<MVVMBinding> ();
		readonly List<MVVMControl> _Controls = new List<MVVMControl> ();
		readonly List<MVVMTrigger> Triggers = new List<MVVMTrigger> ();

		MVVMBinding _Name;
		MVVMBinding _DataContext;
		MVVMControl _Root;
		MVVMValidation _Validation = new MVVMValidation ();

		public MVVMControl Add (MVVMControl control) {
			control._Parent = this;
			Controls.Add (control);
			return this;
		}
		public MVVMControl Add (params MVVMControl[] controls) {
			foreach (MVVMControl control in controls) {
				Add (control);
			}
			return this;
		}
		public MVVMControl Add (MVVMControl control, int index) {
			control._Parent = this;
			Controls.Insert (index, control);
			return this;
		}

		public MVVMControl AddTrigger (MVVMTrigger trigger) {
			trigger.Control = this;
			Triggers.Add (trigger);
			return this;
		}
		public MVVMControl AddTrigger (params MVVMTrigger[] triggers) {
			foreach (MVVMTrigger trigger in triggers) {
				AddTrigger (trigger);
			}
			return this;
		}

		public void RemoveAt (int index) {
			Controls[index].Unbind ();
			Controls.RemoveAt (index);
		}

		public void Swap (int oldIndex, int newIndex) {
			MVVMControl control = Controls[oldIndex];
			Controls[oldIndex] = Controls[newIndex];
			Controls[newIndex] = control;
		}

		public void Move (int oldIndex, int newIndex) {
			MVVMControl control = Controls[oldIndex];
			Controls.RemoveAt (oldIndex);
			Controls.Insert (newIndex, control);
		}

		public void Clear () {
			foreach (MVVMControl control in Controls) {
				control.Unbind ();
			}
			Controls.Clear ();
		}

		public MVVMControl Build (MVVMControl root = null) {
			if (root == null) {
				root = this;
			}
			ControlNames.Clear ();
			ForEach (control => control.Register (root));
			DataContext.Bind ();
			return this;
		}

		public void Unbind () {
			ForEach (control => {
				control.DataContext.Unbind ();
				foreach (MVVMTrigger trigger in control.Triggers) {
					trigger.Unbind ();
				}
				foreach (MVVMBinding binding in control.Bindings) {
					binding.Unbind ();
				}
			});
		}

		void ForEach (Action<MVVMControl> action) {
			action (this);
			foreach (MVVMControl control in Controls) {
				control.ForEach (action);
			}
		}

		void Register (MVVMControl root) {
			_Root = root;
			if (_Name != null) {
				Root.ControlNames[_Name.GetTargetValue<string> ()] = this;
			}
		}

		protected MVVMBinding GetBinding (ref MVVMBinding binding, Action<MVVMBinding> setBinding) {
			if (binding == null) {
				setBinding (new MVVMBinding (string.Empty));
			}
			return binding;
		}

		protected void SetBinding (
			ref MVVMBinding binding, MVVMBinding value,
			MVVMFunc<object> getTargetValue = null, Action<object> setTargetValue = null,
			MVVMAction unbind = null,
			MVVMBindingMode defaultMode = MVVMBindingMode.OneWay,
			MVVMUpdateSourceTrigger defaultUpdateSourceTrigger = MVVMUpdateSourceTrigger.PropertyChanged,
			bool isDataContext = false, [CallerMemberName] string propertyName = null
		) {
			if (!isDataContext && binding != null) {
				Bindings.Remove (binding);
			}
			binding = value;
			if (binding == null) {
				return;
			}
			binding.Control = this;
			if (propertyName == null) {
				propertyName = MVVMAPI.GetCallerMemberName ();
			}
			binding.TargetPropertyName = propertyName;
			binding.IsDataContext = isDataContext;
			binding.OnGetTargetValue = getTargetValue;
			binding.OnSetTargetValue = setTargetValue;
			binding.OnUnbind = unbind;
			binding.DefaultMode = defaultMode;
			binding.DefaultUpdateSourceTrigger = defaultUpdateSourceTrigger;
			if (!isDataContext) {
				Bindings.Add (binding);
			}
		}

		internal protected void OnChanged (MVVMBinding binding, MVVMOnChangedType onChangedType = MVVMOnChangedType.PropertyChanged) {
			if (binding == null || binding.BlockOnChanged || binding.IsTrigger) {
				return;
			}
			switch (binding.GetMode ()) {
				case MVVMBindingMode.TwoWay:
				case MVVMBindingMode.OneWayToSource:
					switch (binding.GetUpdateSourceTrigger ()) {
						case MVVMUpdateSourceTrigger.PropertyChanged:
							if (onChangedType == MVVMOnChangedType.PropertyChanged) {
								binding.OneWaySetSourceValue ();
							}
							break;
						case MVVMUpdateSourceTrigger.LostFocus:
							if (onChangedType == MVVMOnChangedType.LostFocus) {
								binding.OneWaySetSourceValue ();
							}
							break;
					}
					break;
			}
			this.RaisePropertyChanged (binding.TargetPropertyName);
		}

		protected void OnCommand (MVVMBinding command, MVVMBinding parameter) {
			if (command == null) {
				return;
			}
			MVVMRelayCommand relayCommand = command.GetTargetValue<MVVMRelayCommand> ();
			if (relayCommand == null) {
				return;
			}
			object value = parameter == null ? null : parameter.GetTargetValue ();
			if (relayCommand.CanExecute (value)) {
				relayCommand.Execute (value);
			}
		}

		public IEnumerator<MVVMControl> GetEnumerator () {
			return Controls.GetEnumerator ();
		}

		IEnumerator IEnumerable.GetEnumerator () {
			return Controls.GetEnumerator ();
		}

	}

}